collector/agent_poller.go

```diff
@@ -69,80 +61,120 @@ type AgentState struct {
 // This function should be run as a gofunc.
-func RunAgentPoller(recordsChan chan<- *AvroDatum, agentStateChan chan<- *AgentState, stats chan<- StatsEvent) {
+func (a *Agent) Run(recordsChan chan<- *AvroDatum, stats chan<- StatsEvent) {
 	// fetch agent ip once. per DC/OS docs, changing a node IP is unsupported
-	var agentIp string = ""
+	a.AgentIP = ""
 	if len(*agentTestStateFileFlag) == 0 ||
 		len(*agentTestSystemFileFlag) == 0 ||
 		len(*agentTestContainersFileFlag) == 0 {
 		// only get the ip if actually needed
-		agentIp = getAgentIp(stats)
+		//TODO needs err
+		a.getIP(stats)
 	}
 
 	// do one poll immediately upon starting, to ensure that agent metadata is populated early:
-	pollAgent(agentIp, recordsChan, agentStateChan, stats)
-	ticker := time.NewTicker(time.Second * time.Duration(*agentPollPeriodFlag))
+	a.pollAgent(recordsChan, stats)
+	ticker := time.NewTicker(time.Second * time.Duration(a.PollPeriod))
 	for {
 		select {
 		case _ = <-ticker.C:
-			pollAgent(agentIp, recordsChan, agentStateChan, stats)
+			a.pollAgent(recordsChan, stats)
 		}
 	}
 }
 
 // ---
 
-func pollAgent(agentIp string, recordsChan chan<- *AvroDatum, agentStateChan chan<- *AgentState, stats chan<- StatsEvent) {
+func (a *Agent) pollAgent(recordsChan chan<- *AvroDatum, stats chan<- StatsEvent) {
 	// always fetch/emit agent state first: downstream will use it for tagging metrics
-	agentState, err := getAgentState(agentIp, stats)
+	agentState, err := a.getAgentState(stats)
 	if err == nil {
-		agentStateChan <- agentState
+		a.AgentStateChan <- agentState
 	} else {
-		log.Printf("Failed to retrieve state from agent at %s: %s", agentIp, err)
+		log.Printf("Failed to retrieve state from agent at %s: %s", a.AgentIP, err)
 	}
 
-	systemMetricsList, err := getSystemMetrics(agentIp, agentState, stats)
+	systemMetricsList, err := a.getSystemMetrics(agentState, stats)
 	if err == nil {
 		if systemMetricsList != nil {
 			recordsChan <- systemMetricsList
 		}
 	} else {
-		log.Printf("Failed to retrieve system metrics from agent at %s: %s", agentIp, err)
+		log.Printf("Failed to retrieve system metrics from agent at %s: %s", a.AgentIP, err)
 	}
 
-	containerMetricsLists, err := getContainerMetrics(agentIp, agentState, stats)
+	containerMetricsLists, err := a.getContainerMetrics(agentState, stats)
 	if err == nil {
 		for _, metricList := range containerMetricsLists {
 			recordsChan <- metricList
 		}
 	} else {
-		log.Printf("Failed to retrieve container metrics from agent at %s: %s", agentIp, err)
+		log.Printf("Failed to retrieve container metrics from agent at %s: %s", a.AgentIP, err)
 	}
 }
 
 // runs detect_ip => "10.0.3.26\n"
-func getAgentIp(stats chan<- StatsEvent) string {
+func (a *Agent) getIP(stats chan<- StatsEvent) error {
 	stats <- MakeEvent(AgentIpLookup)
-	cmdWithArgs := strings.Split(*agentIpCommandFlag, " ")
+	cmdWithArgs := strings.Split(a.IPCommand, " ")
 	ipBytes, err := exec.Command(cmdWithArgs[0], cmdWithArgs[1:]...).Output()
```

Comment:
https://github.com/dcos/3dt/blob/master/api/helpers.go#L54

^^ I'm going to update it to use this, and remove the IP Command from the config. It's safe to bet this is available to our executable.